### PR TITLE
Take upsert snapshot when creating new consuming segment

### LINKED ISSUE #10800: Upsert snapshot should be taken before a new consuming segment is created
Currently, when snapshot is enabled, snapshot is taken when shutting down the server. This will cause inconsistency because the changes made by the new consuming segment has already been reflected into the snapshots, but when we restart the servers and reload all the segments, the consuming segment will be re-consumed.
This problem is more critical for partial-upsert table because the snapshot will mark the old record as invalid, but the new record still sit in the consuming segment thus not persisted.

The fix should be to persist the snapshot when a new consuming segment is created but before any record is consumed

### PR DESCRIPTION
**Pull request by Jackie-Jiang**
Fix #10800

- Track all the segments managed by the `BasePartitionUpsertMetadataManager`
- Remove the tracking for replaced segments because that can be replaced by tracking all the segments
- Introduce a read-write lock in `BasePartitionUpsertMetadataManager` to ensure segments are not changed when snapshot is taken
- Take snapshot before a new consuming segment starts consuming instead of when the segment is removed (most commonly during server shut down)

Note that it will add some lag (time to persist snapshot) to the consuming segment when it starts consuming, but it is the overhead we need to pay in order to get consistent snapshot.

### CODE REVIEW

**Comment by Jackie-Jiang** [comment]
cc @deemoliu

File: pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java [hunk 1]
```diff
@@ -1309,15 +1310,15 @@ public LLRealtimeSegmentDataManager(SegmentZKMetadata segmentZKMetadata, TableCo
...
     _isReadyToConsumeData = isReadyToConsumeData;
     _segmentVersion = indexLoadingConfig.getSegmentVersion();
     _instanceId = _realtimeTableDataManager.getServerInstance();
     _leaseExtender = SegmentBuildTimeLeaseExtender.getLeaseExtender(_tableNameWithType);
     _protocolHandler = new ServerSegmentCompletionProtocolHandler(_serverMetrics, _tableNameWithType);
     CompletionConfig completionConfig = _tableConfig.getValidationConfig().getCompletionConfig();
-    _segmentCompletionMode = completionConfig != null
-        && CompletionMode.DOWNLOAD.toString().equalsIgnoreCase(completionConfig.getCompletionMode())
-        ? CompletionMode.DOWNLOAD : CompletionMode.DEFAULT;
+    _segmentCompletionMode = completionConfig != null && CompletionMode.DOWNLOAD.toString()
```
Comment (KKcorps) [inline pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java:1320]: nit: Unnecessary formatting

Comment (KKcorps) [inline pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java:1320]: nit: Unnecessary formatting

File: pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java [hunk 2]
```diff
@@ -1360,9 +1361,9 @@ public LLRealtimeSegmentDataManager(SegmentZKMetadata segmentZKMetadata, TableCo
             firstSortedColumn, llcSegmentName);
         sortedColumn = firstSortedColumn;
       } else {
-        _segmentLogger
-            .warn("Sorted column name: {} from RealtimeDataResourceZKMetadata is not existed in schema for segment {}.",
-                firstSortedColumn, llcSegmentName);
+        _segmentLogger.warn(
```
Comment (KKcorps) [inline pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java:1364]: nit: Unnecessary formatting

File: pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java [hunk 3]
```diff
@@ -1386,27 +1387,26 @@ public LLRealtimeSegmentDataManager(SegmentZKMetadata segmentZKMetadata, TableCo
...
     _nullHandlingEnabled = indexingConfig.isNullHandlingEnabled();
 
-    _columnIndicesForRealtimeTable = new ColumnIndicesForRealtimeTable(sortedColumn,
-        new ArrayList<>(indexLoadingConfig.getInvertedIndexColumns()),
-        new ArrayList<>(indexLoadingConfig.getTextIndexColumns()),
-        new ArrayList<>(indexLoadingConfig.getFSTIndexColumns()),
-        new ArrayList<>(indexLoadingConfig.getNoDictionaryColumns()),
-        new ArrayList<>(indexLoadingConfig.getVarLengthDictionaryColumns()));
+    _columnIndicesForRealtimeTable =
+        new ColumnIndicesForRealtimeTable(sortedColumn, new ArrayList<>(indexLoadingConfig.getInvertedIndexColumns()),
```
Comment (KKcorps) [inline pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java:1391]: nit: Unnecessary formatting

File: pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java [hunk 4]
```diff
@@ -1386,27 +1387,26 @@ public LLRealtimeSegmentDataManager(SegmentZKMetadata segmentZKMetadata, TableCo
...
+            .setSegmentName(_segmentNameStr).setStreamName(streamTopic).setSchema(_schema)
+            .setTimeColumnName(timeColumnName).setCapacity(_segmentMaxRowCount)
+            .setAvgNumMultiValues(indexLoadingConfig.getRealtimeAvgMultiValueCount())
+            .setSegmentZKMetadata(segmentZKMetadata).setOffHeap(_isOffHeap).setMemoryManager(_memoryManager)
             .setStatsHistory(realtimeTableDataManager.getStatsHistory())
             .setAggregateMetrics(indexingConfig.isAggregateMetrics())
             .setIngestionAggregationConfigs(IngestionConfigUtils.getAggregationConfigs(tableConfig))
-            .setNullHandlingEnabled(_nullHandlingEnabled)
-            .setConsumerDir(consumerDir).setUpsertMode(tableConfig.getUpsertMode())
+            .setNullHandlingEnabled(_nullHandlingEnabled).setConsumerDir(consumerDir)
```
Comment (KKcorps) [inline pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java:1408]: nit: Unnecessary formatting

File: pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java [hunk 5]
```diff
@@ -1451,17 +1451,16 @@ public LLRealtimeSegmentDataManager(SegmentZKMetadata segmentZKMetadata, TableCo
...
+          llcSegmentName, _segmentMaxRowCount, new DateTime(_consumeEndTime, DateTimeZone.UTC));
       startConsumerThread();
     } catch (Exception e) {
       // In case of exception thrown here, segment goes to ERROR state. Then any attempt to reset the segment from
       // ERROR -> OFFLINE -> CONSUMING via Helix Admin fails because the semaphore is acquired, but not released.
       // Hence releasing the semaphore here to unblock reset operation via Helix Admin.
       _partitionGroupConsumerSemaphore.release();
-      _realtimeTableDataManager.addSegmentError(_segmentNameStr, new SegmentErrorInfo(now(),
-          "Failed to initialize segment data manager", e));
+      _realtimeTableDataManager.addSegmentError(_segmentNameStr,
```
Comment (KKcorps) [inline pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/LLRealtimeSegmentDataManager.java:1462]: nit: Unnecessary formatting

**Comment by KKcorps** [comment]
@Jackie-Jiang the snapshot operation in prod can take till around 30 seconds as per my observation. 

Are we fine with spending that much time between each segment commit?

There will be a momentary consumption lag while snapshot is being taken although I think pinot should be able to easily catch up.

**Comment by Jackie-Jiang** [comment]
> @Jackie-Jiang the snapshot operation in prod can take till around 30 seconds as per my observation.
> Are we fine with spending that much extra time between each segment commit?
> There will be a momentary consumption lag while snapshot is being taken although I think pinot should be able to easily catch up.
> One optimisation we can do is only take snapshot of the segment which were **changed** from the last commit.

@KKcorps Good point. It will definitely increase the delay before a new consuming segment starts consuming events, but might not be the whole 30 seconds because we only persist snapshot for one partition, and it doesn't involve the segment destroy time. I added a TODO to consider optimizing it by only persist the changed snapshot, but want to see how long it usually takes to persist all the snapshot as a baseline.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java [hunk 6]
```diff
@@ -61,8 +63,12 @@ public abstract class BasePartitionUpsertMetadataManager implements PartitionUps
...
   protected final Logger _logger;
 
-  @VisibleForTesting
-  public final Set<IndexSegment> _replacedSegments = ConcurrentHashMap.newKeySet();
+  // Tracks all the segments managed by this manager (excluding EmptySegment)
+  protected final Set<IndexSegment> _trackedSegments = ConcurrentHashMap.newKeySet();
+
+  // NOTE: We do not persist snapshot on the first consuming segment because most segments might not be loaded yet
+  protected volatile boolean _gotFirstConsumingSegment = false;
+  protected final ReadWriteLock _snapshotLock;
```
Comment (walterddr) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java:71]: this lock is used (read-wise) by 3 methods `add/remove/replaceSegment`, therefore should this lock be private?

**Review by walterddr — APPROVED** [review]
lgtm

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java (hunk 6 above)
Comment (walterddr) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java:71]: seems like all readLock are follow by start operation. do we still need the start operation? b/c there couldn't been any concurrent operations with the new lock right?

Comment (walterddr) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/BasePartitionUpsertMetadataManager.java:71]: nvm there's addRecord. my bad

**Review by KKcorps — APPROVED** [review]
LGTM!

**Comment by KKcorps** [comment]
@Jackie-Jiang should I go ahead and merge it?
